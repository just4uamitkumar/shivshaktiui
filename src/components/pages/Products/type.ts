export type ProductType = {
    category:string;
    id:number;
    description?:string;
    price:number;
    images?:string[];
    title:string;
    rating?:number;
    thumbnail:string;
    availabilityStatus:string
    brand:string;
    discountPercentage:number
    shippingInformation:string
}