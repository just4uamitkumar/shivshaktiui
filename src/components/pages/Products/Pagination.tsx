import { Stack } from "@mui/material";
import { useEffect, useState } from "react";
import type { ProductType } from "./type";
import {
  FaRegArrowAltCircleLeft,
  FaRegArrowAltCircleRight,
} from "react-icons/fa";
import IconBtn from "../../common/IconBtn";
import React from "react";

interface Props {
  ProductsList: ProductType[];
  selectPageHandler: (selectedPage: number) => void;
  page: number;
}

const Pagination: React.FC<Props> = ({
  ProductsList,
  selectPageHandler,
  page,
}) => {
  const [pageLength, setPageLength] = useState<number[]>([]);

  useEffect(() => {
    if (ProductsList.length > 0) {
      const n: number = ProductsList.length / 10;
      const pageItem = Array.from({ length: n }, (_, index) => index + 1);
      setPageLength(pageItem);
    }
  }, [ProductsList]);

  console.log("Pagination Component Rendered", {
    ProductsList,
    pageLength,
    page,
  });

  return (
    <Stack
      className="pagination"
      direction={"row"}
      display={"flex"}
      justifyContent={"center"}
      alignItems={"center"}
    >
      <IconBtn
        IconComponent={FaRegArrowAltCircleLeft}
        onClick={() => selectPageHandler(page - 1)}
        iconClass={page > 1 ? "" : "disable"}
      />

      {pageLength &&
        pageLength.map((_, i) => {
          return (
            <span
              key={i}
              className={page === i + 1 ? "activePage" : ""}
              onClick={() => selectPageHandler(i + 1)}
            >
              {i + 1}
            </span>
          );
        })}
      <IconBtn
        IconComponent={FaRegArrowAltCircleRight}
        onClick={() => selectPageHandler(page + 1)}
        iconClass={page < ProductsList?.length / 10 ? "" : "disable"}
      />
    </Stack>
  );
};

export default React.memo(Pagination);
