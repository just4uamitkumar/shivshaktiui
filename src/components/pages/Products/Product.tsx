import { Grid, Stack } from "@mui/material";
import TypoGraphy from "../../common/TypoGraphy";
import CustomBtn from "../../common/Button";
import { getOriginalPrice, navigateUrl } from "../../../utils/commonFunc";
import React from "react";

interface Props {
  id: number;
  title: string;
  thumbnail: string;
  price: number;
  discountPercentage: number;
  category: string;
  brand: string;
  shippingInformation: string;
  availabilityStatus: string;
}

const Product: React.FC<Props> = ({
  id,
  title,
  thumbnail,
  price,
  discountPercentage,
  category,
  brand,
  shippingInformation,
  availabilityStatus,
}) => {
  const getProduct = async (id: number) => {
    navigateUrl(`/Product/${id}`);
  };

  return (
    <>
      <Grid size={4} className="ProductCard" key={id}>
        <Stack className="imgWrapper">
          <img src={thumbnail} alt={title} />
        </Stack>
        <Stack>
          <TypoGraphy variant="h4">{title}</TypoGraphy>
        </Stack>
        <Stack className="price" direction={"row"} display={"flex"}>
          <Stack className="mr-2 buyPrice">
            <TypoGraphy variant="body1" typeClass="semi-bold-font">
              ${price}
            </TypoGraphy>
          </Stack>
          <Stack className="mr-2 line-through">
            <TypoGraphy variant="body1">
              {getOriginalPrice(discountPercentage, price).toFixed(2)}
            </TypoGraphy>
          </Stack>
          <Stack className="green500">
            <TypoGraphy variant="body1" typeClass="orange">
              {`${discountPercentage}% off`}
            </TypoGraphy>
          </Stack>
        </Stack>
        <Stack direction={"row"} display={"flex"}>
          <Stack className="mr-2" direction={"row"}>
            <TypoGraphy variant="body1">
              {"Category : "}
              {category}
            </TypoGraphy>
          </Stack>
          <Stack direction={"row"}>
            <TypoGraphy variant="body1">
              {"Brand : "}
              {brand}
            </TypoGraphy>
          </Stack>
        </Stack>
        <Stack className="shippingInfo">
          <TypoGraphy variant="body1">{shippingInformation}</TypoGraphy>
        </Stack>
        <Stack>
          <TypoGraphy variant="body1">
            <strong>{"Availability : "}</strong>
            {availabilityStatus}
          </TypoGraphy>
        </Stack>
        <Stack justifyContent={"center"}>
          <CustomBtn
            text="Go to Product"
            variant="contained"
            className="primary"
            onClick={() => getProduct(id)}
          />
        </Stack>
      </Grid>
    </>
  );
};

export default React.memo(Product);
