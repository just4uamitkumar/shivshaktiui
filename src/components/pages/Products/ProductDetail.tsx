import { Grid, Stack } from "@mui/material";
import axios from "axios";
import { useEffect, useState } from "react";
import { useParams } from "react-router";
import type { ProductType } from "./type";
import PageBanner from "../../shared/PageBanner";
import TypoGraphy from "../../common/TypoGraphy";
import { getOriginalPrice } from "../../../utils/commonFunc";
import Loader from "../../shared/Loader";
import React from "react";

const ProductDetail: React.FC = () => {
  const { id } = useParams();
  const [data, setData] = useState<ProductType | null>(null);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    if (id) {
      getProduct();
    }
  }, [id]);

  const getProduct = async () => {
    setLoading(true);
    const response = await axios.get(`https://dummyjson.com/products/${id}`);
    setData(response?.data);
    setLoading(false);
  };

  return (
    <>
      <PageBanner title={data ? data.title : "Product Detail"} />
      <Grid size={12} className="container" container>
        <Grid
          size={12}
          className="ProductDetail"
          flexDirection={"row"}
          display={"flex"}
        >
          {loading ? (
            <Loader />
          ) : (
            <>
              <Grid size={4}>
                <Stack className="img-wrap">
                  {data && data.images && data.images.length > 0 && (
                    <img src={data.images[0]} alt={data.title} />
                  )}
                </Stack>
              </Grid>
              <Grid size={8} className="content-wrap">
                <Stack>
                  <TypoGraphy variant="h4">{data?.title}</TypoGraphy>
                </Stack>
                <Stack className="price" direction={"row"} display={"flex"}>
                  <Stack className="mr-2 buyPrice">
                    <TypoGraphy variant="body1" typeClass="semi-bold-font">
                      ${data?.price}
                    </TypoGraphy>
                  </Stack>
                  <Stack className="mr-2 line-through">
                    <TypoGraphy variant="body1">
                      {data &&
                        getOriginalPrice(
                          data?.discountPercentage,
                          data?.price
                        ).toFixed(2)}
                    </TypoGraphy>
                  </Stack>
                  <Stack className="green500">
                    <TypoGraphy variant="body1" typeClass="orange">
                      {`${data?.discountPercentage}% off`}
                    </TypoGraphy>
                  </Stack>
                </Stack>
                <Stack direction={"row"} display={"flex"}>
                  <Stack className="mr-2" direction={"row"}>
                    <TypoGraphy variant="body1">
                      {"Category : "}
                      {data?.category}
                    </TypoGraphy>
                  </Stack>
                  <Stack direction={"row"}>
                    <TypoGraphy variant="body1">
                      {"Brand : "}
                      {data?.brand}
                    </TypoGraphy>
                  </Stack>
                </Stack>
                <Stack className="shippingInfo">
                  <TypoGraphy variant="body1">
                    {data?.shippingInformation}
                  </TypoGraphy>
                </Stack>
                <Stack>
                  <TypoGraphy variant="body1">
                    <strong>{"Availability : "}</strong>
                    {data?.availabilityStatus}
                  </TypoGraphy>
                </Stack>
              </Grid>
            </>
          )}
        </Grid>
      </Grid>
    </>
  );
};

export default React.memo(ProductDetail);
