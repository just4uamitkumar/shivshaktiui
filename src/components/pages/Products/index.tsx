import { Grid, TextField } from "@mui/material";

import PageBanner from "../../shared/PageBanner";
import { useCallback, useEffect, useState } from "react";
import axios from "axios";
import type { ProductType } from "./type";
import Loader from "../../shared/Loader";
import Product from "./Product";
import { useDebounce } from "../../../utils/useDebounce";
import Pagination from "./Pagination";

const Products: React.FC = () => {
  const [ProductsList, setProductsList] = useState<ProductType[]>([]);
  const [filteredProductsList, setFilteredProductsList] = useState<
    ProductType[]
  >([]);
  const [page, setPage] = useState<number>(1);
  const [loading, setLoading] = useState<boolean>(true);
  const [query, setQuery] = useState<string>("");
  const [name, setName] = useState<string>("");
  const [category, setCategory] = useState<string>("");
  const [minPrice, setMinPrice] = useState<string>("");
  const [maxPrice, setMaxPrice] = useState<string>("");
  const debouncedValue = useDebounce(query, 500);
  const debouncedByName = useDebounce(name, 500);
  const debouncedByCategory = useDebounce(category, 500);
  const debouncedByMinPrice = useDebounce(minPrice, 500);
  const debouncedByMaxPrice = useDebounce(maxPrice, 500);

  useEffect(() => {
    getProducts();
  }, []);

  // Get Products by query
  useEffect(() => {
    const getSpecificProducts = async () => {
      setLoading(true);
      if (query !== "") {
        const response = await axios.get(
          `https://dummyjson.com/products/search?q=${query}&limit=100`
        );
        setFilteredProductsList(response?.data?.products);
      } else {
        setFilteredProductsList(ProductsList); // Reset to full data list when search input is empty
      }
      setLoading(false);
    };
    if (debouncedValue) {
      getSpecificProducts();
    }
  }, [debouncedValue, query, ProductsList]);

  // Get Products by name
  useEffect(() => {
    if (debouncedByName && name !== "") {
      const filtered = ProductsList.filter((item) =>
        item.title.toLowerCase().includes(name.toLowerCase())
      );
      setFilteredProductsList(filtered);
    } else {
      setFilteredProductsList(ProductsList); // Reset to full data list when search input is empty
    }
  }, [name, ProductsList, debouncedByName]);

  // Get Products by category
  useEffect(() => {
    if (debouncedByCategory && category !== "") {
      const filtered = ProductsList.filter((item) =>
        item.category.toLowerCase().includes(category.toLowerCase())
      );
      setFilteredProductsList(filtered);
    } else {
      setFilteredProductsList(ProductsList); // Reset to full data list when search input is empty
    }
  }, [category, ProductsList, debouncedByCategory]);

  // Get Products by min /max price
  useEffect(() => {
    if (
      debouncedByMinPrice &&
      debouncedByMaxPrice &&
      minPrice !== "" &&
      maxPrice !== ""
    ) {
      const filtered = ProductsList.filter(
        (item) =>
          item.price >= parseFloat(minPrice) &&
          item.price <= parseFloat(maxPrice)
      );
      setFilteredProductsList(filtered);
    } else if (debouncedByMinPrice && minPrice !== "" && maxPrice === "") {
      const filtered = ProductsList.filter(
        (item) => item.price >= parseFloat(minPrice)
      );
      setFilteredProductsList(filtered);
    } else if (debouncedByMaxPrice && maxPrice !== "" && minPrice === "") {
      const filtered = ProductsList.filter(
        (item) => item.price <= parseFloat(maxPrice)
      );
      setFilteredProductsList(filtered);
    } else {
      setFilteredProductsList(ProductsList); // Reset to full data list when search input is empty
    }
  }, [
    minPrice,
    maxPrice,
    ProductsList,
    debouncedByMinPrice,
    debouncedByMaxPrice,
  ]);

  const getProducts = async () => {
    setLoading(true);
    const response = await axios.get(
      "https://dummyjson.com/products?limit=200"
    );
    setProductsList(response?.data?.products);
    setFilteredProductsList(response?.data?.products);
    setLoading(false);
  };

  const selectPageHandler = useCallback(
    (selectedPage: number) => {
      if (
        selectedPage >= 1 &&
        selectedPage <= ProductsList?.length / 10 &&
        selectedPage !== page
      ) {
        setPage(selectedPage);
      }
    },
    [ProductsList?.length, page]
  );

  return (
    <>
      <PageBanner title={"Products"} />

      <Grid className="container pb-4 pt-4" spacing={3} container>
        <Grid size={12} display={"flex"} flexWrap={"wrap"} gap={2}>
          <TextField
            value={query}
            variant="outlined"
            placeholder="Search products..."
            type="text"
            onChange={(
              e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
            ) => setQuery(e.target.value)}
          />
          <TextField
            value={name}
            variant="outlined"
            type="text"
            placeholder="Search by name..."
            onChange={(
              e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
            ) => setName(e.target.value)}
          />
          <TextField
            value={category}
            variant="outlined"
            type="text"
            placeholder="Search by category..."
            onChange={(
              e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
            ) => setCategory(e.target.value)}
          />
          <TextField
            value={minPrice}
            variant="outlined"
            type="text"
            placeholder="Search by Min Price"
            onChange={(
              e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
            ) => setMinPrice(e.target.value)}
          />
          <TextField
            value={maxPrice}
            variant="outlined"
            type="text"
            placeholder="Search by Max Price"
            onChange={(
              e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
            ) => setMaxPrice(e.target.value)}
          />
        </Grid>
        {loading ? (
          <Loader />
        ) : (
          filteredProductsList?.length > 0 &&
          filteredProductsList
            ?.slice(page * 10 - 10, page * 10)
            ?.map((item: ProductType) => (
              <Product
                key={item?.id}
                id={item?.id}
                title={item?.title}
                thumbnail={item.thumbnail}
                price={item.price}
                discountPercentage={item.discountPercentage}
                category={item.category}
                brand={item.brand}
                shippingInformation={item.shippingInformation}
                availabilityStatus={item.availabilityStatus}
              />
            ))
        )}
        {filteredProductsList?.length > 0 && (
          <Grid size={12}>
            <Pagination
              ProductsList={filteredProductsList}
              selectPageHandler={selectPageHandler}
              page={page}
            />
          </Grid>
        )}
      </Grid>
    </>
  );
};

export default Products;
