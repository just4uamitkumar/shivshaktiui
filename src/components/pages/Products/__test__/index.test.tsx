import { describe, it, expect, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";

import Products from "../index";
import axios from "axios";

// Mock axios
vi.mock("axios");
const mockedAxios = axios as unknown as { get: ReturnType<typeof vi.fn> };

const mockProducts = [
  {
    id: 1,
    title: "Mock Product 1",
    price: 100,
    thumbnail: "mock1.jpg",
    category: "Category A",
    brand: "BrandA",
    discountPercentage: 5,
    shippingInformation: "Fast",
    availabilityStatus: "In Stock",
  },
  {
    id: 2,
    title: "Mock Product 2",
    price: 200,
    thumbnail: "mock2.jpg",
    category: "Category B",
    brand: "BrandB",
    discountPercentage: 10,
    shippingInformation: "Standard",
    availabilityStatus: "Out of Stock",
  },
];

describe("Products Component", () => {
  it("matches snapshot", () => {
    const { container } = render(<Products />);
    expect(container).toMatchSnapshot();
  });

  it("should render without crashing", () => {
    render(<Products />);
    expect(Products).toBeDefined();
  });
  it("should render the page title", () => {
    const title = "Products";
    render(<Products />);
    const pageTitle = screen.getByText(title);
    expect(pageTitle).toBeInTheDocument();
  });

  it("fetches and displays products", async () => {
    // Set up mock response
    mockedAxios.get.mockResolvedValueOnce({
      data: { products: mockProducts },
    });

    render(<Products />);

    await waitFor(() => {
      expect(screen.getByText("Mock Product 1")).toBeInTheDocument();
      expect(screen.getByText("Mock Product 2")).toBeInTheDocument();
    });
  });

//   it("shows loader initially", () => {
//     render(<Products />);
//     expect(screen.getByTestId("loader")).toBeInTheDocument();
//   });
});
