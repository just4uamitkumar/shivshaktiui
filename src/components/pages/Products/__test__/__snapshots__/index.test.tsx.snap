// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Products Component > matches snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row pageBanner css-d64g0g-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-direction-xs-row container css-17fpwt7-MuiGrid-root"
    >
      <div
        class="MuiStack-root pageTitle css-nen11g-MuiStack-root"
      >
        Products
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-3 container pb-4 pt-4 css-1u8xl8z-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-direction-xs-row MuiGrid-grid-xs-12 css-1yceax4-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r0»"
            placeholder="Search products..."
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r1»"
            placeholder="Search by name..."
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r2»"
            placeholder="Search by category..."
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r3»"
            placeholder="Search by Min Price"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r4»"
            placeholder="Search by Max Price"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="loader"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
        role="progressbar"
        style="width: 3rem; height: 3rem;"
      >
        <svg
          class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
